src/main/java/carpet/commands/DrawCommand.java

```diff
@@ -144,21 +235,269 @@ private static int drawCircle(ServerCommandSource source, BlockPos pos, double r
+            Predicate<CachedBlockPosition> replacement, boolean solid) {
+        int affected = 0;
+
+        BlockPos.Mutable mbpos = new BlockPos.Mutable(pos);
+
+        List<BlockPos> list = Lists.newArrayList();
+
+        ServerWorld world = source.getWorld();
+
+        CarpetSettings.impendingFillSkipUpdates = !CarpetSettings.fillUpdates;
+
+        for (int x = -radius; x <= radius; ++x) {
+
+            for (int y = -radius; y <= radius; ++y) {
+
+                for (int z = -radius; z <= radius; ++z) {
+
+                    if (manhattan(0, 0, 0, x, y, z) != radius) {
+                        continue;
+                    }
+
+                    mbpos.set(pos.getX() + x, pos.getY() + y, pos.getZ() + z);
+
+                    if (replacement == null || replacement.test(new CachedBlockPosition(world, mbpos, true))) {
+                        BlockEntity tileentity = world.getBlockEntity(mbpos);
+
+                        if (tileentity instanceof Inventory) {
+                            ((Inventory) tileentity).clear();
+                        }
+
+                        if (block.setBlockState(world, mbpos, 2)) {
+                            list.add(mbpos.toImmutable());
+                            ++affected;
+                        }
+                    }
+                }
+            }
+        }
+
+        CarpetSettings.impendingFillSkipUpdates = false;
+
+        if (CarpetSettings.fillUpdates) {
+
+            for (BlockPos blockpos1 : list) {
+                Block blokc = world.getBlockState(blockpos1).getBlock();
+                world.updateNeighbors(blockpos1, blokc);
+            }
+        }
+
+        Messenger.m(source, "gi Filled " + affected + " blocks");
+
+        return 1;
+    }
+
+    private static int drawConey(ServerCommandSource source, BlockPos pos, int radius, int height, boolean pointup, BlockStateArgument block,
+    Predicate<CachedBlockPosition> replacement) {
+        return (int) drawConey(source, pos, radius, height, block, replacement, pointup, false);
+    }
+
+    private static double drawConey(ServerCommandSource source, BlockPos pos, int radius, int height,BlockStateArgument block,
+            Predicate<CachedBlockPosition> replacement, boolean pointup, boolean solid) {
+        int affected = 0;
+        int offset = 0;
+        if(pointup){
+            offset=height-1;
+        }
+        BlockPos.Mutable mbpos = new BlockPos.Mutable(pos);
+
+        List<BlockPos> list = Lists.newArrayList();
+
+        ServerWorld world = source.getWorld();
+
+        CarpetSettings.impendingFillSkipUpdates = !CarpetSettings.fillUpdates;
+
+        for (int x = -radius; x <= radius; ++x) {
+
+            for (int y = -offset; y <= height-offset-1; ++y) {
+
+                for (int z = -radius; z <= radius; ++z) {
+
+                    if (!(x*x+z*z<=(radius/height*y)*(radius/height*y))) {
```

Comment:
Same for cone - this is not the way to iterate over blocks in a cone.

Reply:
then how?

Reply:
shall I remove?

Reply:
also, this is a lot more efficient than the diamond iterator, as unlike the diamond which only fills in 11% of iterated blocks, the cone fills 26%. then again you could fill in the diamond and that gives 18% efficeincy (Here i calculate efficiency by the number of blocks filled divided by the number of blocks iterated over, as that gives a pretty good estimate of how many blocks are being 'wasted', so to speak)

Reply:
and the new diamond iterator only ever iterates on blocks it needs to iterate on, so by this calculation, it is 100% efficient

Reply:
also, I'll have to rework the cone thing, as it currently only properly works if the base and height are the same, cos otherwise they dont return integers. I think it's as simple as just rounding, but ill try nontheless

Reply:
done: see fe0eccc